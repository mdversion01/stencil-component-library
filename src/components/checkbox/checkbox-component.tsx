import { Component, Prop, h, State, Event, EventEmitter } from '@stencil/core';

@Component({
  tag: 'checkbox-component',
  styleUrls: [
    './checkbox.scss',
    '../form-styles.scss',
    '../custom-form-inputs.scss'
  ],
  shadow: false,
})
export class CheckboxComponent {
  @Prop() checkbox = false;
  @Prop() checkboxGroup = false;
  @Prop() customCheckbox = false;
  @Prop() customCheckboxGroup = false;

  @Prop() groupOptions: any = []; // JSON string or array
  @Prop() groupTitle = '';
  @Prop() groupTitleSize = '';
  @Prop() inline = false;
  @Prop() inputId = '';
  @Prop() name = '';
  @Prop() noPadding = false;
  @Prop() labelTxt = '';
  @Prop() required = false;
  @Prop() size = '';
  @Prop() checked: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() validation = false;
  @Prop() validationMsg = '';
  @Prop() value = '';

  @State() parsedOptions: Array<any> = [];
  @State() checkedValues: string[] = [];
  @State() singleChecked = false;

  @Event() groupChange: EventEmitter<string[]>;

  componentWillLoad() {
    if (Array.isArray(this.groupOptions)) {
      this.parsedOptions = this.groupOptions;
    } else if (typeof this.groupOptions === 'string') {
      try {
        this.parsedOptions = JSON.parse(this.groupOptions);
      } catch (e) {
        console.error('Invalid groupOptions JSON:', e);
        this.parsedOptions = [];
      }
    }

    this.checkedValues = this.parsedOptions.filter(opt => opt.checked).map(opt => opt.value);
  }

  private handleGroupChange(event: Event, value: string) {
    const target = event.target as HTMLInputElement;
    let updated = [...this.checkedValues];

    if (target.checked) {
      if (!updated.includes(value)) {
        updated.push(value);
      }
    } else {
      updated = updated.filter(v => v !== value);
    }

    this.checkedValues = updated;
    this.groupChange.emit(this.checkedValues);
  }

  private handleSingleChange(event: Event) {
    const target = event.target as HTMLInputElement;
    this.singleChecked = target.checked;
  }

  render() {
    // GROUP RENDER
    if (this.checkboxGroup || this.customCheckboxGroup) {
      const wrapperClass = this.customCheckboxGroup ? 'custom-control custom-checkbox' : 'form-check';
      const inputClass = this.customCheckboxGroup ? 'custom-control-input' : 'form-check-input';
      const labelClass = this.customCheckboxGroup ? 'custom-control-label' : 'form-check-label';

      const showValidation = this.validation && this.required && this.checkedValues.length === 0;

      return (
        <div class={`checkbox-group ${showValidation ? 'was-validated' : ''}`}>
          <div class={`group-title ${this.groupTitleSize}`}>
            {this.groupTitle}
            {this.required ? <span class="required">*</span> : ''}
          </div>
          <div class={`form-group ${this.inline ? 'form-inline' : ''} no-pad`}>
            {this.parsedOptions.map(option => (
              <div class={wrapperClass}>
                <input
                  class={`${inputClass} ${this.size}`}
                  type="checkbox"
                  name={this.name}
                  id={option.inputId}
                  value={option.value}
                  checked={this.checkedValues.includes(option.value)}
                  disabled={option.disabled}
                  required={this.required}
                  aria-checked={this.checkedValues.includes(option.value) ? 'true' : 'false'}
                  aria-disabled={option.disabled ? 'true' : 'false'}
                  onChange={e => this.handleGroupChange(e, option.value)}
                />
                <label class={`${labelClass} ${this.size}`} htmlFor={option.inputId}>
                  {option.labelTxt}
                </label>
              </div>
            ))}
          </div>
          {showValidation && this.validationMsg ? (
            <div class="invalid-feedback form-text" style={{ display: showValidation ? 'block' : '' }}>
              {this.validationMsg}
            </div>
          ) : null}
        </div>
      );
    }

    // SINGLE RENDER
    const singleWrapperClass = this.customCheckbox ? 'custom-control custom-checkbox' : 'form-check';
    const singleInputClass = this.customCheckbox ? 'custom-control-input' : 'form-check-input';
    const singleLabelClass = this.customCheckbox ? 'custom-control-label' : 'form-check-label';

    const showSingleValidation = this.validation && this.required && !this.singleChecked;

    return (
      <div class={`form-group check-box no-pad ${showSingleValidation ? 'was-validated' : ''}`}>
        <div class={singleWrapperClass}>
          <input
            class={`${singleInputClass} ${this.size}`}
            id={this.inputId}
            type="checkbox"
            name={this.name}
            value={this.value}
            disabled={this.disabled}
            required={this.required}
            aria-checked={this.singleChecked ? 'true' : 'false'}
            aria-disabled={this.disabled ? 'true' : 'false'}
            onChange={e => this.handleSingleChange(e)}
          />
          <label class={`${singleLabelClass} ${this.size}`} htmlFor={this.inputId}>
            {this.labelTxt}
            {this.required ? <span class="required">*</span> : ''}
          </label>
          {showSingleValidation && this.validationMsg ? <div class="invalid-feedback form-text">{this.validationMsg}</div> : null}
        </div>
      </div>
    );
  }
}
