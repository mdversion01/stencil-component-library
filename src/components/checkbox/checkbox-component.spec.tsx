import { newSpecPage } from '@stencil/core/testing';
import { CheckboxComponent } from './checkbox-component';

describe('checkbox-component', () => {
  it('renders single standard checkbox', async () => {
    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component checkbox input-id="check1" name="test" label-txt="Accept terms"></checkbox-component>`,
    });
    const input = page.root.querySelector('input');
    expect(input.getAttribute('type')).toBe('checkbox');
    expect(input.getAttribute('name')).toBe('test');
    expect(input.getAttribute('id')).toBe('check1');
    expect(page.root).toMatchSnapshot();
  });

  it('renders single custom checkbox', async () => {
    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component custom-checkbox input-id="check2" name="custom" label-txt="Custom Option"></checkbox-component>`,
    });
    expect(page.root.querySelector('.custom-control')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('renders checkbox group with options', async () => {
    const groupOptions = JSON.stringify([
      { value: 'opt1', labelTxt: 'Option 1', inputId: 'opt1', disabled: false, checked: true },
      { value: 'opt2', labelTxt: 'Option 2', inputId: 'opt2', disabled: false, checked: false },
    ]);

    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component checkbox-group name="group1" group-title="Choose Options" group-options='${groupOptions}' required validation validation-msg="Please select at least one"></checkbox-component>`,
    });

    const inputs = page.root.querySelectorAll('input[type="checkbox"]');
    expect(inputs.length).toBe(2);
    expect(inputs[0].checked).toBe(true);
    expect(inputs[1].checked).toBe(false);
    expect(page.root).toMatchSnapshot();
  });

  it('renders custom checkbox group with options', async () => {
    const groupOptions = JSON.stringify([
      { value: 'optA', labelTxt: 'Custom A', inputId: 'optA', disabled: false, checked: false },
      { value: 'optB', labelTxt: 'Custom B', inputId: 'optB', disabled: true, checked: false },
    ]);

    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component custom-checkbox-group name="group2" group-title="Custom Group" group-options='${groupOptions}'></checkbox-component>`,
    });

    const inputs = page.root.querySelectorAll('input[type="checkbox"]');
    expect(inputs.length).toBe(2);
    expect(inputs[1].disabled).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('shows validation message when no group option is selected and required', async () => {
    const groupOptions = JSON.stringify([
      { value: 'val1', labelTxt: 'Val 1', inputId: 'val1', disabled: false, checked: false },
      { value: 'val2', labelTxt: 'Val 2', inputId: 'val2', disabled: false, checked: false },
    ]);

    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component checkbox-group name="group3" group-title="Validation Test" group-options='${groupOptions}' required validation validation-msg="Please select an option"></checkbox-component>`,
    });

    expect(page.root.querySelector('.invalid-feedback')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('removes validation message when at least one group option is checked', async () => {
    const groupOptions = JSON.stringify([
      { value: 'val1', labelTxt: 'Val 1', inputId: 'val1', disabled: false, checked: false },
      { value: 'val2', labelTxt: 'Val 2', inputId: 'val2', disabled: false, checked: false },
    ]);

    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component checkbox-group name="group4" group-title="Validation Clear" group-options='${groupOptions}' required validation validation-msg="Select one"></checkbox-component>`,
    });

    const secondInput = page.root.querySelectorAll('input')[1];
    secondInput.checked = true;
    secondInput.dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(page.root.querySelector('.invalid-feedback')).toBeFalsy();
    expect(page.root).toMatchSnapshot();
  });

  it('emits groupChange event on group change', async () => {
    const groupOptions = JSON.stringify([
      { value: 'x', labelTxt: 'X', inputId: 'x', disabled: false },
      { value: 'y', labelTxt: 'Y', inputId: 'y', disabled: false },
    ]);

    const page = await newSpecPage({
      components: [CheckboxComponent],
      html: `<checkbox-component checkbox-group name="group5" group-title="Emit Test" group-options='${groupOptions}'></checkbox-component>`,
    });

    const spy = jest.fn();
    page.root.addEventListener('groupChange', spy);

    const secondInput = page.root.querySelectorAll('input')[1];
    secondInput.checked = true;
    secondInput.dispatchEvent(new Event('change'));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalled();
  });
});
