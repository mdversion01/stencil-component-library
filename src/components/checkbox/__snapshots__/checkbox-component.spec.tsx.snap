// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox-component removes validation message when at least one group option is checked 1`] = `
<checkbox-component checkbox-group="" group-options="[{&quot;value&quot;:&quot;val1&quot;,&quot;labelTxt&quot;:&quot;Val 1&quot;,&quot;inputId&quot;:&quot;val1&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false},{&quot;value&quot;:&quot;val2&quot;,&quot;labelTxt&quot;:&quot;Val 2&quot;,&quot;inputId&quot;:&quot;val2&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false}]" group-title="Validation Clear" name="group4" required="" validation="" validation-msg="Select one">
  <div class="checkbox-group">
    <div class="group-title">
      Validation Clear
      <span class="required">
        *
      </span>
    </div>
    <div class="form-group no-pad">
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="val1" name="group4" required="" type="checkbox" value="val1">
        <label class="form-check-label" htmlfor="val1">
          Val 1
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="true" aria-disabled="false" checked="" class="form-check-input" id="val2" name="group4" required="" type="checkbox" value="val2">
        <label class="form-check-label" htmlfor="val2">
          Val 2
        </label>
      </div>
    </div>
  </div>
</checkbox-component>
`;

exports[`checkbox-component renders checkbox group with options 1`] = `
<checkbox-component checkbox-group="" group-options="[{&quot;value&quot;:&quot;opt1&quot;,&quot;labelTxt&quot;:&quot;Option 1&quot;,&quot;inputId&quot;:&quot;opt1&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:true},{&quot;value&quot;:&quot;opt2&quot;,&quot;labelTxt&quot;:&quot;Option 2&quot;,&quot;inputId&quot;:&quot;opt2&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false}]" group-title="Choose Options" name="group1" required="" validation="" validation-msg="Please select at least one">
  <div class="checkbox-group">
    <div class="group-title">
      Choose Options
      <span class="required">
        *
      </span>
    </div>
    <div class="form-group no-pad">
      <div class="form-check">
        <input aria-checked="true" aria-disabled="false" checked="" class="form-check-input" id="opt1" name="group1" required="" type="checkbox" value="opt1">
        <label class="form-check-label" htmlfor="opt1">
          Option 1
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="opt2" name="group1" required="" type="checkbox" value="opt2">
        <label class="form-check-label" htmlfor="opt2">
          Option 2
        </label>
      </div>
    </div>
  </div>
</checkbox-component>
`;

exports[`checkbox-component renders custom checkbox group with options 1`] = `
<checkbox-component custom-checkbox-group="" group-options="[{&quot;value&quot;:&quot;optA&quot;,&quot;labelTxt&quot;:&quot;Custom A&quot;,&quot;inputId&quot;:&quot;optA&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false},{&quot;value&quot;:&quot;optB&quot;,&quot;labelTxt&quot;:&quot;Custom B&quot;,&quot;inputId&quot;:&quot;optB&quot;,&quot;disabled&quot;:true,&quot;checked&quot;:false}]" group-title="Custom Group" name="group2">
  <div class="checkbox-group">
    <div class="group-title">
      Custom Group
    </div>
    <div class="form-group no-pad">
      <div class="custom-checkbox custom-control">
        <input aria-checked="false" aria-disabled="false" class="custom-control-input" id="optA" name="group2" type="checkbox" value="optA">
        <label class="custom-control-label" htmlfor="optA">
          Custom A
        </label>
      </div>
      <div class="custom-checkbox custom-control">
        <input aria-checked="false" aria-disabled="true" class="custom-control-input" disabled="" id="optB" name="group2" type="checkbox" value="optB">
        <label class="custom-control-label" htmlfor="optB">
          Custom B
        </label>
      </div>
    </div>
  </div>
</checkbox-component>
`;

exports[`checkbox-component renders single custom checkbox 1`] = `
<checkbox-component custom-checkbox="" input-id="check2" label-txt="Custom Option" name="custom">
  <div class="check-box form-group no-pad">
    <div class="custom-checkbox custom-control">
      <input aria-checked="false" aria-disabled="false" class="custom-control-input" id="check2" name="custom" type="checkbox" value="">
      <label class="custom-control-label" htmlfor="check2">
        Custom Option
      </label>
    </div>
  </div>
</checkbox-component>
`;

exports[`checkbox-component renders single standard checkbox 1`] = `
<checkbox-component checkbox="" input-id="check1" label-txt="Accept terms" name="test">
  <div class="check-box form-group no-pad">
    <div class="form-check">
      <input aria-checked="false" aria-disabled="false" class="form-check-input" id="check1" name="test" type="checkbox" value="">
      <label class="form-check-label" htmlfor="check1">
        Accept terms
      </label>
    </div>
  </div>
</checkbox-component>
`;

exports[`checkbox-component shows validation message when no group option is selected and required 1`] = `
<checkbox-component checkbox-group="" group-options="[{&quot;value&quot;:&quot;val1&quot;,&quot;labelTxt&quot;:&quot;Val 1&quot;,&quot;inputId&quot;:&quot;val1&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false},{&quot;value&quot;:&quot;val2&quot;,&quot;labelTxt&quot;:&quot;Val 2&quot;,&quot;inputId&quot;:&quot;val2&quot;,&quot;disabled&quot;:false,&quot;checked&quot;:false}]" group-title="Validation Test" name="group3" required="" validation="" validation-msg="Please select an option">
  <div class="checkbox-group was-validated">
    <div class="group-title">
      Validation Test
      <span class="required">
        *
      </span>
    </div>
    <div class="form-group no-pad">
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="val1" name="group3" required="" type="checkbox" value="val1">
        <label class="form-check-label" htmlfor="val1">
          Val 1
        </label>
      </div>
      <div class="form-check">
        <input aria-checked="false" aria-disabled="false" class="form-check-input" id="val2" name="group3" required="" type="checkbox" value="val2">
        <label class="form-check-label" htmlfor="val2">
          Val 2
        </label>
      </div>
    </div>
    <div class="form-text invalid-feedback" style="display: block;">
      Please select an option
    </div>
  </div>
</checkbox-component>
`;
